### PR TITLE
895 Update temp directory function in all tests and examples

### LINKED ISSUE #895: Update all tests to use tempfile.TemporaryDirectory
**Is your feature request related to a problem? Please describe.**
Currently, all the unit tests use `tempfile.mkdtemp` to create temp dir, as @ericspod suggested, it's better to use `tempfile.TemporaryDirectory` because cleanup is guaranteed even if the block within the context raises an exception.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #895 .

### Description
This PR updated the `tempfile` usage in all the tests and examples according to @ericspod 's suggestion.
I didn't change the usage in integration tests because it's used as a class member and didn't change the usage in all the notebooks because it's hard to use "with ... as ... :" in notebooks.

And I also fixed several missing "delete" issue for temp directory in tests, thanks for @ericspod 's suggestion, `tempfile.TemporaryDirectory` is really better to avoid errors.

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by ericspod** [comment]
I'm good with this, does anyone have any other input @wyli @atbenmurray ?

**Comment by wyli** [comment]
this was changed back and forth @Nic-Ma  and now I can't remember why we did those, e.g.
https://github.com/Project-MONAI/MONAI/pull/446
https://github.com/Project-MONAI/MONAI/commit/10203f32c4707a2cd11e49262252b14a59aec52a

**Comment by Nic-Ma** [comment]
Hi @wyli ,

The previous 2 PRs were:
1. At the beginning, only few tests used `tempdir` and others used current dir, so the first PR is to use `tempdir` for all.
2. The second PR is because I thought `with tempfile.TemporaryDirectory() as tempdir:` is using system `/tmp/` folder, so I used `tempfile. mkdtemp()` to replace them, which is not correct.

Now thanks for @ericspod 's suggestion, `with tempfile.TemporaryDirectory() as tempdir:` actually creates a temp dir and good for our cases, I think it's much more clear now.

Thanks.

File: examples/segmentation_3d/unet_evaluation_array.py [hunk 1]
```diff
@@ -33,58 +32,57 @@ def main():
     config.print_config()
     logging.basicConfig(stream=sys.stdout, level=logging.INFO)
 
-    tempdir = tempfile.mkdtemp()
-    print(f"generating synthetic data to {tempdir} (this may take a while)")
-    for i in range(5):
-        im, seg = create_test_image_3d(128, 128, 128, num_seg_classes=1)
+    with tempfile.TemporaryDirectory() as tempdir:
```
Comment (wyli) [inline examples/segmentation_3d/unet_evaluation_array.py:35]: I think we should avoid this large block of indentation by adding the context from outside of main, something like:
```python
 if __name__ == "__main__":
    with tempfile.TemporaryDirectory() as tempdir:
        main(tempdir)
```

Comment (Nic-Ma) [reply examples/segmentation_3d/unet_evaluation_array.py:35]: good point, I will change it soon.
Thanks.

**Pull request merged**